adding inference trace injection (#36890)
Azure/azure-sdk-for-python: sdk/ai/azure-ai-inference/README.md

To generate embeddings for additional phrases, simply call `client.embed` multiple times using the same `client`.
-->

> ## Tracing

Reviewer: I would put this section at the very bottom of the readme, since this is not really a "quickstart" kind of thing, but advanced.

Author: moved after troubleshooting

The Azure AI Inferencing API Tracing library provides tracing for Azure AI Inference client library for Python. Refer to Installation chapter above for installation instructions.

### Setup

> The environment variable AZURE_TRACING_GEN_AI_CONTENT_RECORDING_ENABLED controls whether the actual message contents will be included in the traces or not. By default, the message contents are not include as part of the trace. Set the value of the environment variable to true (case insensitive) for the message contents to be included as part of the trace. Any other value will cause the message contents not to be traced.

Reviewer: Quotes around the "true" in ".. variable to true (case insensitive)"

Author: fixed

> The environment variable AZURE_TRACING_GEN_AI_CONTENT_RECORDING_ENABLED controls whether the actual message contents will be included in the traces or not. By default, the message contents are not include as part of the trace. Set the value of the environment variable to true (case insensitive) for the message contents to be included as part of the trace. Any other value will cause the message contents not to be traced.

Reviewer 2: ```suggestion
The environment variable AZURE_TRACING_GEN_AI_CONTENT_RECORDING_ENABLED controls whether the actual message contents will be included in the traces or not. By default, the message contents are not recorded. Set the value of the environment variable to true (case insensitive) for the message contents to be included as part of the trace. Any other value will cause the message contents not to be traced.
```

Author: fixed

> You also need to configure the tracing implementation in your code, like so:

Reviewer 2: ```suggestion
You also need to configure the tracing implementation in your code by setting `AZURE_SDK_TRACING_IMPLEMENTATION` to `opentelemetry` or configuring it in the code with the following  snippet:
```

Author: fixed

```
from azure.core.settings import settings
settings.tracing_implementation = "opentelemetry"
```

> 

Reviewer 2: ```suggestion
Please refer to [azure-core-tracing-documentation](https://learn.microsoft.com/python/api/overview/azure/core-tracing-opentelemetry-readme)
```

It should be updated soon - I updated readme recently - https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core-tracing-opentelemetry/README.md

Author: fixed

> ### Trace Exporter(s)

Reviewer 2: I don't think you really need this section - just  refer to the https://learn.microsoft.com/python/api/overview/azure/core-tracing-opentelemetry-readme

Author: shortened the chapter and added the link

In order for the traces to be captured, you need to setup the applicable trace exporters. The chosen exporter will be based on where you want the traces to be output. You can also implement your own exporter. The first example below shows how to setup an exporter to Azure Monitor.

> Please refer to [this](https://learn.microsoft.com/en-us/azure/azure-monitor/app/create-workspace-resource?tabs=bicep) documentation for more information about how to created Azure Monitor resource.

Reviewer: remove "en-us" from link

Author: fixed

> Please refer to [this](https://learn.microsoft.com/en-us/azure/azure-monitor/app/create-workspace-resource?tabs=bicep) documentation for more information about how to created Azure Monitor resource.

Reviewer 3: created -> create

Author: fixed

> Configure the APPLICATIONINSIGHTS_CONNECTION_STRING based on your Azure Monitor resource.

Reviewer: Configure the environment variable ...

Author: fixed

```

> # Setup tracing to Azure Monitor

Reviewer 2: You don't really need this section - just link https://learn.microsoft.com/python/api/overview/azure/core-tracing-opentelemetry-readme (it will soon be auto-updated with much easier way documented here - https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core-tracing-opentelemetry/README.md)

The resulting code snippet is just 
```python
pip install azure-monitor-opentelemetry
configure_azure_monitor(
   connection_string="<your-connection-string>"
)
```

Author: shortened the chapter and added the link

> from azure.monitor.opentelemetry.exporter import AzureMonitorLogExporter

Reviewer: Please look into automatically injecting these code snippets from your sample, so we are able to run them. 

See wiki page: https://github.com/Azure/azure-sdk-for-python/blob/main/doc/dev/code_snippets.md

To get the code snippets inserted, open a command prompt windows in the sdk\ai\azure-ai-inference folder, andrun this:
`python ..\..\..\tools\azure-sdk-tools\ci_tools\snippet_update\python_snippet_updater.py .`
You will need to run it twice to make the error go away.

tracer = trace.get_tracer(__name__)
span_processor = BatchSpanProcessor(
    AzureMonitorTraceExporter.from_connection_string(
        os.environ["APPLICATIONINSIGHTS_CONNECTION_STRING"]
    )
)
trace.get_tracer_provider().add_span_processor(span_processor)
```

The following example shows how to setup tracing to console output.

```
# Setup tracing to console
exporter = ConsoleSpanExporter()
trace.set_tracer_provider(TracerProvider())
tracer = trace.get_tracer(__name__)
trace.get_tracer_provider().add_span_processor(SimpleSpanProcessor(exporter))
```
### Instrumentation

Use the AIInferenceInstrumentor to instrument the Azure AI Inferencing API for LLM tracing, this will cause the LLM traces to be emitted from Azure AI Inferencing API.

```
from azure.core.tracing import AIInferenceApiInstrumentor
# Instrument AI Inference API
AIInferenceApiInstrumentor().instrument()
```

It is also possible to uninstrument the Azure AI Inferencing API by using the uninstrument call. After this call, the LLM traces will no longer be emitted by the Azure AI Inferencing API until instrument is called again.

```
AIInferenceApiInstrumentor().uninstrument()
```

### Tracing Your Own Functions

> The @tracer.start_as_current_span decorator can be used to trace your own functions. This will trace the function parameters and their values. You can also add further attributes to the span in the function implementation as demonstrated below. Note that you will have to setup the tracer in your code before using the decorator.

Reviewer 2: Please add a link to https://opentelemetry.io/docs/languages/python/ so that users know where to find more info

Author: added

```

> # The @tracer.start_as_current_span decorator will

Reviewer 2: please add tracer initialization here

```suggestion
from opentelemetry.trace import get_tracer
tracer = get_tracer(__name__)

# The @tracer.start_as_current_span decorator will
```

Author: added

* Have a look at the [Samples](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/ai/azure-ai-inference/samples) folder, containing fully runnable Python code for doing inference using synchronous and asynchronous clients.

> ## Tracing

Reviewer: I think this order of sections makes more sense:

## Troubleshooting
## Tracing
## Next steps
## Contributing

Author: okay.  i had it like that at one point, but then changed it before committing the change

### Setup

The environment variable AZURE_TRACING_GEN_AI_CONTENT_RECORDING_ENABLED controls whether the actual message contents will be recorded in the traces or not. By default, the message contents are not recorded as part of the trace. When message content recording is disabled any function call tool related function names, function parameter names and function parameter values are also not recorded in the trace. Set the value of the environment variable to "true" (case insensitive) for the message contents to be recorded as part of the trace. Any other value will cause the message contents not to be recorded.

You also need to configure the tracing implementation in your code by setting `AZURE_SDK_TRACING_IMPLEMENTATION` to `opentelemetry` or configuring it in the code with the following snippet:

<!-- SNIPPET:sample_chat_completions_with_tracing.trace_setting -->

```python
from azure.core.settings import settings
settings.tracing_implementation = "opentelemetry"
```

<!-- END SNIPPET -->

Please refer to [azure-core-tracing-documentation](https://learn.microsoft.com/python/api/overview/azure/core-tracing-opentelemetry-readme) for more information.

> ### Trace Exporter(s)

Reviewer 2: ```suggestion
### Exporting traces with OpenTelemetry
```

> In order for the traces to be captured, you need to setup the applicable trace exporters. The chosen exporter will be based on where you want the traces to be output. You can also implement your own exporter. You can find more information [here](https://learn.microsoft.com/en-us/python/api/overview/azure/core-tracing-opentelemetry-readme?view=azure-python-preview).

Reviewer 2: ```suggestion
Azure AI Inference is instrumented with OpenTelemetry. In order to enable tracing you need to configure OpenTelemetry to export traces to your observability backend. 

Refer to [Azure SDK tracing in Python](https://learn.microsoft.com/python/api/overview/azure/core-tracing-opentelemetry-readme?view=azure-python-preview) for more details.
```

Author: changed

> Please refer to [this](https://learn.microsoft.com/en-us/azure/azure-monitor/app/create-workspace-resource?tabs=bicep) documentation for more information about how to create Azure Monitor resource for the Azure Monitor exporter.

Reviewer 2: ```suggestion
Refer to [Azure Monitor OpenTelemetry documentation](https://learn.microsoft.com/azure/azure-monitor/app/opentelemetry-enable?tabs=python) for the details on how to send Azure AI Inference traces to Azure Monitor and create Azure Monitor resource.
```

Author: changed
